storage/innobase/fil/fil0crypt.cc

```diff
@@ -1711,6 +2111,47 @@ static bool fil_crypt_find_space_to_rotate(key_state_t *key_state,
   return false;
 }
 
+bool fil_space_crypt_t::re_encrypt_validation_tag(const uint from_key_version,
+                                                  const uint to_key_version) {
+  if (from_key_version > to_key_version)
+    return true;  // re-encryption not needed
+
+  // load key_versions that we might be missing in the cache
+  if (load_keys_to_local_cache(from_key_version, to_key_version) == false) {
+    // generate error
+    return false;
+  }
+
+  byte re_encrypted_validation_tag[ENCRYPTION_KEYRING_VALIDATION_TAG_SIZE] = {
+      0};
+  byte copy_encrypted_validation_tag[ENCRYPTION_KEYRING_VALIDATION_TAG_SIZE];
+  memcpy(copy_encrypted_validation_tag, encrypted_validation_tag,
+         ENCRYPTION_KEYRING_VALIDATION_TAG_SIZE);
+
+  for (uint key_version = from_key_version; key_version <= to_key_version;
+       ++key_version) {
```

Comment:
Do we write intermediate versions of the tag to page 0? Also, if the write to page 0 is still in page 0, we need to validate using the version found from redo [we discussed on the channel]

Reply:
We don't write intermediate versions of the tag to page0. Added validation if correct encryption keys are available when processing redo log entry.